durango-cli/src/main.rs

```diff
@@ -0,0 +1,275 @@
+
+use structopt::StructOpt;
+use std::path;
+use env_logger;
+use log::{info, warn, error};
+use ring;
+use hex;
+use std::process;
+use durango::Durango;
+use std::fs;
+use std::io;
+use std::io::Read;
+use std::io::Write;
+
+
+#[derive(Debug, StructOpt)]
+#[structopt(rename_all="kebab")]
+struct Opt {
+    /// Path to policy file
+    #[structopt(parse(from_os_str))]
+    policy_path: path::PathBuf,
+
+    /// Path to client certificate file
+    #[structopt(short, long, parse(from_os_str))]
+    identity: path::PathBuf,
+
+    /// Path to client key file
+    #[structopt(short, long, parse(from_os_str))]
+    key: path::PathBuf,
+
+    /// Specify optional program file to upload
+    ///
+    /// Accepts "-" to read from stdin
+    ///
+    /// Note: This requires "PiProvider" permissions in the
+    /// policy file.
+    #[structopt(short, long, parse(from_os_str))]
+    program: Option<path::PathBuf>,
+
+    /// Specify optional data file to upload
+    ///
+    /// Accepts "-" to read from stdin
+    ///
+    /// Note: This requires "DataProvider" permissions in the
+    /// policy file.
+    #[structopt(short, long, multiple=true, number_of_values=1, parse(from_os_str))]
+    data: Vec<path::PathBuf>,
+
+    /// Specify optional output file to store results. If not provided
+    /// the results will not be fetched.
+    ///
+    /// If --no-shutdown is not provided, Durango will request a shutdown
+    /// from the Sinaloa server after recieving the results.
+    ///
+    /// Accepts "-" to write to stdout
+    ///
+    /// Note: This requires "ResultReader" permissions in the
+    /// policy file.
+    #[structopt(short, long, visible_alias="result", parse(from_os_str))]
+    output: Option<path::PathBuf>,
+
+    /// Do not request a shutdown of the Sinaloa server after recieving the
+    /// results. This can be useful if you have multiple result readers.
+    #[structopt(short, long)]
+    no_shutdown: bool,
+
+    /// Request shutdown without requesting data.
+    ///
+    /// Note: This requires "ResultReader" permissions in the
+    /// policy file.
+    #[structopt(short, long)]
+    shutdown: bool,
+}
+
+
+/// Entry point
+fn main() {
+    // parse args
+    let opt = Opt::from_args();
+
+    // setup logger
+    // TODO, unlike sinaloa/tabasco, this is a client, do we really
+    // need timestamps/context?
+    env_logger::from_env(
+        env_logger::Env::default().default_filter_or("info")
+    ).init();
+
+    // load policy
+    info!("Loading policy {:?}", opt.policy_path);
+    let policy_json = match std::fs::read_to_string(&opt.policy_path) {
+        Ok(policy_json) => policy_json,
+        Err(_) => {
+            error!("Cannot open file {:?}", opt.policy_path);
+            process::exit(1);
+        }
+    };
+    let policy_hash_bytes = ring::digest::digest(
+        &ring::digest::SHA256, policy_json.as_bytes());
+    let policy_hash = hex::encode(&policy_hash_bytes.as_ref().to_vec());
```

Comment:
Can this be simplified to `policy_hash_bytes.as_ref()`?

Reply:
i.e. without converting to a vector then taking a reference.

Reply:
Good point, will update